packages/amplify_api/example/ios/unit_tests/FlutterURLSessionTests.swift

```diff
@@ -0,0 +1,263 @@
+//
+//  FlutterURLSessionTests.swift
+//  unit_tests
+//
+//  Created by Nys, Dillon on 7/13/21.
+//
+
+import Foundation
+import XCTest
+import Combine
+@testable import Amplify
+import AmplifyPlugins
+@testable import amplify_api
+
+typealias MockObserver = PassthroughSubject<Never, Never>
+
+struct HTTPStatusResponse: Codable {
+    let code: Int
+    let description: String
+}
+
+class MockSessionDelegate: URLSessionBehaviorDelegate {
+    let subject: MockObserver
+    
+    init(subject: MockObserver) {
+        self.subject = subject
+    }
+    
+    func urlSessionBehavior(_ session: URLSessionBehavior, dataTaskBehavior: URLSessionDataTaskBehavior, didCompleteWithError error: Error?) {
+        subject.send(completion: .finished)
+    }
+    
+    func urlSessionBehavior(_ session: URLSessionBehavior, dataTaskBehavior: URLSessionDataTaskBehavior, didReceive data: Data) {}
+}
+
+class FlutterURLSessionTests: XCTestCase {
+    static let factory = FlutterURLSessionBehaviorFactory()
+    var cancellables = Set<AnyCancellable>()
+    
+    let testBody = "{\"test\":\"test\"}"
+    let httpStatusHeaders = [ "Accept": "application/json" ]
+    
+    var awsPlugin: AWSAPIPlugin!
+    
+    override static func setUp() {
+        Amplify.reset()
+    }
+
+    override static func tearDown() {
+        Amplify.reset()
+    }
+
+    override func setUp() {
+        cancellables = []
+    }
+    
+    override func setUpWithError() throws {
+        Amplify.reset()
+        awsPlugin = AWSAPIPlugin(sessionFactory: FlutterURLSessionTests.factory)
+        let config = """
+        {
+            "UserAgent": "aws-amplify-cli/2.0",
+            "Version": "1.0",
+            "api": {
+                "plugins": {
+                    "awsAPIPlugin": {
+                        "mockAPI": {
+                            "endpointType": "REST",
+                            "endpoint": "https://httpstat.us",
+                            "region": "us-west-2",
+                            "authorizationType": "NONE"
+                        }
+                    }
+                }
+            }
+        }
+        """.data(using: .utf8)!
+        let amplifyConfig = try JSONDecoder().decode(AmplifyConfiguration.self, from: config)
+        try Amplify.add(plugin: awsPlugin)
+        try Amplify.configure(amplifyConfig)
+    }
+    
+    private func makeOperation() -> RESTOperation {
+        MockRESTOperation(request: RESTRequest(), operationType: .get, result: .empty)
+    }
+    
+    private func willThrow(_ statusCode: Int) -> Bool { !(200 ..< 300).contains(statusCode) }
+    
+    private func runMockTest(statusCode: Int, method: RESTOperationType, body: String? = nil) {
+        let cancelToken = "\(statusCode).\(method.rawValue)"
+        let operation = makeOperation()
+        OperationsManager.addOperation(cancelToken: cancelToken, operation: operation)
+        
+        let url = URL(string: "https://httpstat.us/\(statusCode)")!
```

Comment:
Good points. These would technically be integration tests. Actual unit testing would require mocking of URLSession, which I'm not sure is worth the effort. If there is a better place to put these tests, though, let me know. 

I agree that depending on a 3rd party service is not ideal. Fortunately it would be pretty easy to spin up a lambda which does the same thing - let me know if that's a better approach and I'll be happy to do that.

Reply:
Thanks for explanation. If too difficult to mock, can just leave this way. We can just keep it in mind. Adding another lambda would mean keeping it the correct account for long-term support, may not be as stable.